src/spark_manager.py

```diff
@@ -143,16 +151,229 @@ def create_cluster(self,
+                        }
+                    },
+                    "spec": {
+                        "hostname": f"spark-master-{self.username.lower()}",
+                        "containers": [
+                            {
+                                "name": "spark-master",
+                                "image": self.image,
+                                "imagePullPolicy": self.image_pull_policy,
+                                "ports": [
+                                    {
+                                        "containerPort": self.DEFAULT_MASTER_PORT,
+                                        "name": "master-comm",
+                                        "protocol": "TCP",
+                                    },
+                                    {
+                                        "containerPort": self.DEFAULT_MASTER_WEBUI_PORT,
+                                        "name": "master-ui",
+                                        "protocol": "TCP",
+                                    },
+                                ],
+                                "env": env,
+                                "resources": {
+                                    "requests": {"memory": memory, "cpu": str(cores)}
+                                    # TODO: Add resource limits if needed
+                                },
+                            }
+                        ],
+                    },
+                },
+            },
+        }
+
+        try:
+            # Try to create the deployment first
+            self.apps_api.create_namespaced_deployment(
+                namespace=self.namespace, body=deployment
+            )
+            logger.info(f"Created Spark master deployment: {self.master_name}")
+        except ApiException as e:
+            if e.status == 409:  # Conflict - already exists
+                try:
+                    # Delete the existing deployment
+                    self.apps_api.delete_namespaced_deployment(
+                        name=self.master_name, namespace=self.namespace
+                    )
+                    logger.info(
+                        f"Deleted existing Spark master deployment: {self.master_name}"
+                    )
+
+                    # Create new deployment
+                    self.apps_api.create_namespaced_deployment(
+                        namespace=self.namespace, body=deployment
+                    )
+                    logger.info(
+                        f"Recreated Spark master deployment: {self.master_name}"
+                    )
+                except ApiException as delete_error:
+                    logger.error(
+                        f"Error replacing Spark master deployment: {delete_error}"
+                    )
+                    raise
+            else:
+                logger.error(f"Error creating Spark master deployment: {e}")
+                raise
+
+    def _create_master_service(self):
```

Comment:
I'm not sure I totally understand what happens when creating a deployment vs. creating a service, could you elaborate?

Reply:
A Deployment primarily manages the lifecycle of Pods, while a Service handles networking and access to those Pods.

Reply:
:+1: